develop/dev-guide-sample-application-java-jdbc.md

```diff
@@ -161,6 +161,31 @@ cd tidb-java-jdbc-quickstart
     make
     ```
 
+    若运行 `make` 时出现错误：
```

Comment:
可以麻烦提供一下运行错误的环境吗？比如，OS、版本号等。

Reply:
OS: Ubuntu
Version: 20,04.6 LTS
sh: ohmyzsh

Reply:
我曾在 Ubuntu + dash，及 MacOS + zsh 上测试通过。是否是 oh-my-zsh 的配置问题？

Reply:
❯ make
bash env.sh && make build run
make[1]: Entering directory '/root/TiDB/tidb-java-jdbc-quickstart'
mvn clean package
[INFO] Scanning for projects...
[INFO]
[INFO] --------------------< com.pingcap:plain-java-jdbc >---------------------
[INFO] Building plain-java-jdbc 0.0.1
[INFO] --------------------------------[ jar ]---------------------------------
[WARNING] The artifact mysql:mysql-connector-java:jar:8.0.33 has been relocated to com.mysql:mysql-connector-j:jar:8.0.33: MySQL Connector/J artifacts moved to reverse-DNS compliant Maven 2+ coordinates.
[INFO]
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ plain-java-jdbc ---
[INFO] Deleting /root/TiDB/tidb-java-jdbc-quickstart/target
[INFO]
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ plain-java-jdbc ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /root/TiDB/tidb-java-jdbc-quickstart/src/main/resources
[INFO]
[INFO] --- maven-compiler-plugin:3.1:compile (default-compile) @ plain-java-jdbc ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 3 source files to /root/TiDB/tidb-java-jdbc-quickstart/target/classes
[INFO]
[INFO] --- maven-resources-plugin:2.6:testResources (default-testResources) @ plain-java-jdbc ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /root/TiDB/tidb-java-jdbc-quickstart/src/test/resources
[INFO]
[INFO] --- maven-compiler-plugin:3.1:testCompile (default-testCompile) @ plain-java-jdbc ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 1 source file to /root/TiDB/tidb-java-jdbc-quickstart/target/test-classes
[INFO]
[INFO] --- maven-surefire-plugin:2.12.4:test (default-test) @ plain-java-jdbc ---
[INFO] Surefire report directory: /root/TiDB/tidb-java-jdbc-quickstart/target/surefire-reports

-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running com.pingcap.AppTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.067 sec

Results :

Tests run: 1, Failures: 0, Errors: 0, Skipped: 0

[INFO]
[INFO] --- maven-jar-plugin:2.4:jar (default-jar) @ plain-java-jdbc ---
[INFO] Building jar: /root/TiDB/tidb-java-jdbc-quickstart/target/plain-java-jdbc.jar
[INFO]
[INFO] --- maven-assembly-plugin:3.3.0:single (make-assembly) @ plain-java-jdbc ---
[INFO] Building jar: /root/TiDB/tidb-java-jdbc-quickstart/target/plain-java-jdbc-jar-with-dependencies.jar
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  5.421 s
[INFO] Finished at: 2024-02-02T11:51:21+08:00
[INFO] ------------------------------------------------------------------------
java -jar target/plain-java-jdbc-jar-with-dependencies.jar
PlayerDAO ERROR: { state => 08S01, cause => com.mysql.cj.exceptions.CJCommunicationsException: Communications link failure

The last packet sent successfully to the server was 0 milliseconds ago. The driver has not received any packets from the server., message => Communications link failure

The last packet sent successfully to the server was 0 milliseconds ago. The driver has not received any packets from the server. }
[createPlayers] ERROR: { state => 08S01, cause => com.mysql.cj.exceptions.CJCommunicationsException: Communications link failure

The last packet sent successfully to the server was 0 milliseconds ago. The driver has not received any packets from the server., message => Communications link failure

The last packet sent successfully to the server was 0 milliseconds ago. The driver has not received any packets from the server. }
com.mysql.cj.jdbc.exceptions.CommunicationsException: Communications link failure

The last packet sent successfully to the server was 0 milliseconds ago. The driver has not received any packets from the server.
        at com.mysql.cj.jdbc.exceptions.SQLError.createCommunicationsException(SQLError.java:175)
        at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:64)
        at com.mysql.cj.jdbc.ConnectionImpl.createNewIO(ConnectionImpl.java:825)
        at com.mysql.cj.jdbc.ConnectionImpl.<init>(ConnectionImpl.java:446)
        at com.mysql.cj.jdbc.ConnectionImpl.getInstance(ConnectionImpl.java:239)
        at com.mysql.cj.jdbc.NonRegisteringDriver.connect(NonRegisteringDriver.java:188)
        at com.mysql.cj.jdbc.MysqlDataSource.getConnection(MysqlDataSource.java:417)
        at com.mysql.cj.jdbc.MysqlDataSource.getConnection(MysqlDataSource.java:129)
        at com.mysql.cj.jdbc.MysqlDataSource.getConnection(MysqlDataSource.java:114)
        at com.pingcap.PlayerDAO.createPlayers(PlayerDAO.java:67)
        at com.pingcap.JDBCExample.main(JDBCExample.java:53)
Caused by: com.mysql.cj.exceptions.CJCommunicationsException: Communications link failure

The last packet sent successfully to the server was 0 milliseconds ago. The driver has not received any packets from the server.
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
        at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:480)
        at com.mysql.cj.exceptions.ExceptionFactory.createException(ExceptionFactory.java:62)
        at com.mysql.cj.exceptions.ExceptionFactory.createException(ExceptionFactory.java:105)
        at com.mysql.cj.exceptions.ExceptionFactory.createException(ExceptionFactory.java:150)
        at com.mysql.cj.exceptions.ExceptionFactory.createCommunicationsException(ExceptionFactory.java:166)
        at com.mysql.cj.protocol.a.NativeSocketConnection.connect(NativeSocketConnection.java:89)
        at com.mysql.cj.NativeSession.connect(NativeSession.java:121)
        at com.mysql.cj.jdbc.ConnectionImpl.connectOneTryOnly(ConnectionImpl.java:945)
        at com.mysql.cj.jdbc.ConnectionImpl.createNewIO(ConnectionImpl.java:815)
        ... 8 more
Caused by: java.net.ConnectException: Connection refused
        at java.base/sun.nio.ch.Net.connect0(Native Method)
        at java.base/sun.nio.ch.Net.connect(Net.java:579)
        at java.base/sun.nio.ch.Net.connect(Net.java:568)
        at java.base/sun.nio.ch.NioSocketImpl.connect(NioSocketImpl.java:593)
        at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:327)
        at java.base/java.net.Socket.connect(Socket.java:633)
        at com.mysql.cj.protocol.StandardSocketFactory.connect(StandardSocketFactory.java:153)
        at com.mysql.cj.protocol.a.NativeSocketConnection.connect(NativeSocketConnection.java:63)
        ... 11 more
PlayerDAO.getPlayer ERROR: { state => 08S01, cause => com.mysql.cj.exceptions.CJCommunicationsException: Communications link failure

The last packet sent successfully to the server was 0 milliseconds ago. The driver has not received any packets from the server., message => Communications link failure

The last packet sent successfully to the server was 0 milliseconds ago. The driver has not received any packets from the server. }
Exception in thread "main" java.lang.NullPointerException: Cannot invoke "com.pingcap.PlayerBean.getId()" because "testPlayer" is null
        at com.pingcap.JDBCExample.main(JDBCExample.java:58)
make[1]: *** [Makefile:24: run] Error 1
make[1]: Leaving directory '/root/TiDB/tidb-java-jdbc-quickstart'
make: *** [Makefile:18: all] Error 2

Reply:
❯ sh env.sh
env.sh: 14: [: true: unexpected operator

Reply:
❯ dash env.sh
env.sh: 14: [: true: unexpected operator

Reply:
我看了下，主要是在Makefile文件这里。
all:
	. env.sh
这里我用make执行时，报错

. env.sh && make build run
/bin/sh: 1: .: env.sh: not found
make: *** [Makefile:18: all] Error 2

Reply:
把makefile的 . env.sh 换成  bash env.sh.
env.sh 的#!/bin/sh  换成 #!/bin/bash
就可以了。

Reply:
好像是make会默认强制用sh，但我记得ubuntu默认用的是bash吧。